### ISSUE TITLE
poll/select for sockets

### ISSUE THREAD

**Issue by StillWaters77** [emscripten-core/emscripten#19447]
This is more of a general question. Since epoll is not supported are poll and select supported? The website states that: "The following POSIX sockets functions are currently not proxied (and will not work): poll(), close() (use shutdown() instead), select()". Is this statement still true? And if so, how can i wait for events on sockets and filedescriptors at the same time?

**Comment by sbc100** [comment]
I don't think we have any way to do that in emscripten today.   In the general case, it is not possible to perform any kind of blocking operation on the web so anything that involves synchronously waiting is unlikely to work.    We do have `-sASYNCIFY` as a workaround but IIRC nobody has yet tried to build select or poll based on `ASYNCIFY`.

**Comment by StillWaters77** [comment]
Since my knowledge of web development is limited, maybe you could explain to me if this idea that i have to implement a select/poll could work:
Basically we threat every fd we want to wait on (in the webcase maybe only websocket/emscripten proxy socket) as a Promise. A read promise gets resolved, e.g. if we received something on the websocket (asynchronously) and a close promise gets resolved, if the connection was closed. This should be implementable since there are already callbacks for websockets in emscripten. A poll/select would than be nothing more than a Promise.any() with all the fds (promises) and a certain timeout promise.
Once an event (or timeout) happens we resolve the promise from the outside "unblocking" the Promise.any and setting a flag in the filedescriptor set for the activated promise.

**Comment by StillWaters77** [comment]
Yes thank you. I implemented the described solution above really primitively and it works like a charm. Only thing is i can not use the proxying of tcp over websockets as the APIs dont offer the same callbacks. Is it possible to register the same callbacks for the tcp connections as for the websockets? E.g. so on socket read i can resolve the specific read promise and on close the close promise?

**Comment by hedwigz** [comment]
@StillWaters77 can you share code? I am trying to achieve something similar
